src/components/atoms/IconInput/IconInput.tsx

```diff
@@ -0,0 +1,57 @@
+import React, { ChangeEvent } from "react";
+import classNames from "classnames";
+import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
+import { IconProp } from "@fortawesome/fontawesome-svg-core";
+import "./IconInput.scss";
+
+export interface IconInputProps {
+  className?: string;
+  icon?: IconProp;
+  onChange?: (e?: ChangeEvent<HTMLInputElement>) => void;
+  value?: string;
+  autoComplete?: string;
+  placeholder?: string;
+  iconPosition?: "right" | "left";
+}
+
+const ICON_INPUT_CLASS_NAME = "icon-input";
+const ICON_CLASS_NAME = `${ICON_INPUT_CLASS_NAME}__icon`;
+const INPUT_CLASS_NAME = `${ICON_INPUT_CLASS_NAME}__input`;
```

Comment:
While I understand and see what you were going for here, I feel like this extra 'indirection' here makes it the code harder to read (and search for) than just manually constructing the strings below. Since they'll only really be passed into the couple of `classNames` calls, I don't feel like the duplication/etc will be a big deal. It's not like we're scattering them across the codebase.

That will likely also make it easier/more likely for IDE/etc static code checkers to correctly be able to hint on these classes/allow to click through to their definitions, etc.

```suggestion
```

Reply:
> this extra 'indirection' here makes it the code harder to read (and search for)

yeah, you're right. I'll rewrite that.

<hr />

btw, if we want to use BEM class names, we can consider this package which generates class names in the needed format
https://www.npmjs.com/package/b_

or do you think it'll be better / easier to build them manually?

Reply:
Since that link seemed to break GitHub's autolinker: [https://github.com/azproduction/b_](https://github.com/azproduction/b_)

Interesting concept.. i've never ever thought to use something like that before tbh.

Also came across https://github.com/ybiquitous/bem-ts which seems to have been updated more recently/might be better maintained.

---

I think these still probably don't add a huge amount of gain/productivity, and might break IDE features/make searching for the classes harder, etc. Given we'd also need to separately write the BEM parts in the SCSS too, we're probably still going to have to think about the 'structure' of it, etc.

There appear to be some eslint/stylelint related plugins for detecting/enforcing BEM patterns, my plans were to enable them at some point to make it a less manual process of 'ensuring we get it right'.

So i'd say probably not worth adding these libs here.